agent/src/main/java/com/thoughtworks/go/agent/AgentHTTPClientController.java

```diff
@@ -105,9 +111,13 @@ public void execute() {
     @Override
     public void work() {
         LOG.debug("[Agent Loop] Trying to retrieve work.");
-        retrieveCookieIfNecessary();
-        retrieveWork();
-        LOG.debug("[Agent Loop] Successfully retrieved work.");
+        if (pluginJarLocationMonitor.getLastRun() > 0) {
```

Comment:
Maybe wrap around a [`CountDownLatch`](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CountDownLatch.html) instead?

Reply:
Yes, that makes more sense. I'll make the change.

Reply:
I tried using a CountDownLatch, but it was becoming difficult to update the flaky tests (`DefaultPluginJarLocationMonitorTest` and `DefaultExternalPluginJarLocationMonitorTest`). That's because some of the tests expect the monitor thread to be run twice - once for a setup and then for an update. The `CountDownLatch.await()` method waits until the coundown has reached zero. So I would have to reset the countdown each time the monitor thread is run. Doing that from the run loop of the monitor thread could again cause the tests to become flaky. Or the other way could be add a startSignal and downSignal as two countDownLatches to synchronize the tests. 

The `lastRun` time (or any increasing counter) seems to be easier to use.